core/src/main/java/org/infinispan/remoting/inboundhandler/NonTotalOrderPerCacheInboundInvocationHandler.java

```diff
@@ -27,15 +57,44 @@ public void handle(CacheRpcCommand command, Reply reply, DeliverOrder order) {
       }
       try {
          final int commandTopologyId = extractCommandTopologyId(command);
-         final boolean waitTransactionalData = command.getCommandId() != StateRequestCommand.COMMAND_ID;
          final boolean onExecutorService = executeOnExecutorService(order, command);
-         handleRunnable(createDefaultRunnable(command, reply, commandTopologyId, waitTransactionalData, onExecutorService), onExecutorService);
+         final BlockingRunnable runnable;
+
+         switch (command.getCommandId()) {
+            case SingleRpcCommand.COMMAND_ID:
+               //lock ownership needs the first topology. we need to check it before invoking the getLockPromise()
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
```

Comment:
TBH I don't really like the approach of keeping the knowledge of all the stages in a single class , I was thinking more of a chain approach, like with `CompletableFuture`. I realize `CompletableFuture` is not a perfect fit here, because you also need a way to propagate cancellation, but still I think it would be good to have more uniformity between the pending lock promises, the real lock promises, and the eventual topology promises.